plugins/http_plugin/http_plugin.cpp

```diff
@@ -202,320 +60,25 @@ class http_plugin_impl : public std::enable_shared_from_this<http_plugin_impl> {
-          * @return abstract_conn_ptr backed by type specific implementations of the methods
-          */
-         template<typename T>
-         static detail::abstract_conn_ptr make_abstract_conn_ptr( detail::connection_ptr<T> conn, http_plugin_impl_ptr impl ) {
-            return std::make_shared<abstract_conn_impl<T>>(std::move(conn), std::move(impl));
-         }
-
-         /**
-          * Helper type that wraps an object of type T and records its "in flight" size to
-          * http_plugin_impl::bytes_in_flight using RAII semantics
-          *
-          * @tparam T - the contained Type
-          */
-         template<typename T>
-         struct in_flight {
-            in_flight(T&& object, http_plugin_impl_ptr impl)
-            :_object(std::move(object))
-            ,_impl(std::move(impl))
-            {
-               _count = detail::in_flight_sizeof(_object);
-               _impl->bytes_in_flight += _count;
-            }
-
-            ~in_flight() {
-               if (_count) {
-                  _impl->bytes_in_flight -= _count;
-               }
-            }
-
-            // No copy constructor, but allow move
-            in_flight(const in_flight&) = delete;
-            in_flight(in_flight&& from)
-            :_object(std::move(from._object))
-            ,_count(from._count)
-            ,_impl(std::move(from._impl))
-            {
-               from._count = 0;
-            }
-
-            // No copy assignment, but allow move
-            in_flight& operator=(const in_flight&) = delete;
-            in_flight& operator=(in_flight&& from) {
-               _object = std::move(from._object);
-               _count = from._count;
-               _impl = std::move(from._impl);
-               from._count = 0;
-            }
-
-            /**
-             * const accessor
-             * @return const reference to the contained object
-             */
-            const T& obj() const {
-               return _object;
-            }
-
-            /**
-             * mutable accessor (can be moved from)
-             * @return mutable reference to the contained object
-             */
-            T& obj() {
-               return _object;
-            }
-
-            T _object;
-            size_t _count;
-            http_plugin_impl_ptr _impl;
-         };
-
-         /**
-          * convenient wrapper to make an in_flight<T>
-          */
-         template<typename T>
-         static auto make_in_flight(T&& object, http_plugin_impl_ptr impl) {
-            return std::make_shared<in_flight<T>>(std::forward<T>(object), std::move(impl));
-         }
+         shared_ptr<beast_http_listener<plain_session, tcp, tcp_socket_t > >  beast_server;
+         shared_ptr<beast_http_listener<ssl_session, tcp, tcp_socket_t > >  beast_https_server;
+         shared_ptr<beast_http_listener<unix_socket_session, stream_protocol, stream_protocol::socket > > beast_unix_server;
```

Comment:
No reason for `beast_server`, `beast_https_server`, or `beast_unix_server` to be `shared_ptr`, they can be `unique_ptr`.

Reply:
beast_http_listener calls async methods (async_accept) and so needs to be shared_from_this(), which implies they need to be constructed through a shared_ptr

Reply:
Ah, missed that, thanks.

Reply:
Calling async methods doesn't require shared_ptrs, it just requires that the object remains "alive" until completion. The shared_from_this() pattern is one way of accomplishing that by dragging the shared_ptr along as a captured variable in each callback. But that pattern tends to be best for ephemeral connection objects imo, because any alternative (that I'm aware of) tends to be clunky.

In this particular case, it does seem like from an architectural standpoint there wouldn't be a need for them to be shared_ptrs. As long as the thread pool is stopped & joined before the unique_ptr goes out of scope everything is safe because the listener is always alive while any asyncs are outstanding.

Also not sure why thread_pool has been promoted to a shared_ptr. shared_ptrs tend to cloud ownership semantics (the other comment about who is to call stop() on the thread_pool being a perfect example); I'm somewhat skeptical they are needed beyond usage of the shared_from_this() pattern on the connection objects.

Reply:
@spoonincode  
On the surface, it seems that the listener objects should only live as long as the http_plugin object, which should "own" them.   However, this is not necessary the case precisely because the listener/session objects are calling async methods, and the http_plugin object lives inside the main app thread, while the listener/session objects live inside the thread_pool.  

Because we never restart the http_plugin while the app is running, having the http_plugin object own the listener object outright, would be less of a problem.  However, on shutdown, because they live in separate threads, the listener objects can very well have outstanding async calls after they would be destroyed by http_plugin object's destructor. In this case, it is likely a segfault or  or an unhandled excpetion error will occur, causing in the terminate_scenarios_XXX and restart_scenarios_XXX to become flaky.